src/v/storage/log_manager.h

```diff
@@ -159,14 +161,19 @@ struct log_config {
  * where each core manages a distinct set of logs. When the service is
  * shut down, calling `stop` on the log manager will close all of the
  * logs currently being managed.
+ *
+ * The storage::api reference may be nullptr (e.g. in tests). It is used to
+ * provide a way for the storage layer to wake up the space management control
+ * loop with lower latency than normal monitoring frequency.
  */
 class log_manager {
 public:
     explicit log_manager(
       log_config,
       kvstore& kvstore,
       storage_resources&,
-      ss::sharded<features::feature_table>&) noexcept;
+      ss::sharded<features::feature_table>&,
+      ss::sharded<storage::api>*) noexcept;
```

Comment:
Circular reference feels awkward.

If the control loop for space monitoring was kept in `disk_space_manager` then this could just be a reference to that object instead of the child having a reference to its parent.

Reply:
Yep--this was needed (or at least something like it) when blocking produce calls because I anticipated that we would want very reactive behavior to avoid blocking, and the way to achieve that would be to have the accounting parts of the storage layer be able to poke the control loop.

But now that we aren't blocking produce given the other discussion in this PR, I think that is much less important now.